### PR TITLE
Automatic differentiation for the loss function.

### PR DESCRIPTION
**Pull request by Sopel97**
This is a WIP PR for replacing the machinery around the loss function with a more robust approach based on automatic differentiation. The advantages are twofold: 1. We don't have to compute the loss derivatives. 2. Loss always matches its derivative.

Currently I allowed myself to remove all the old loss code and replace it with cross entropy based loss. That means `use_wdl` and teacher score rescaling have no effect right now. I would like to bring back at least the `use_wdl` option, but I'd have to understand it deeper and create necessary automatic differentiation primitives to represent stockfish's wdl function with it.

As it stands, I trained one network with this trainer and it turned out, as expected, about equal to the network trained with an older version (Baseline was halfka flip with unoptimized master that's ~20% slower. The net trained with this PR turned out to be ~30 elo better, which accounting for all the things seems about right. Definitely proves correctness).

The design of this automatic differentiation framework relies on constructing expression templates of the computation tree. Each node contains two methods - one for computing the value, and one for computing the gradient. This is a very simple approach and only handles the univariate case with scalar inputs and outputs, but it's enough for any loss calculation we could imagine. Two optimizations are performed with this expression template scheme:
1. Memoization of the computed results, as it's common that reevaluation happens as we don't have precise control over common subexpression elimination.
2. Nodes passed as lvalue references to a node are assumed to outlive said node. Nodes passed as rvalue references are moved to storage inside the node. This has a big effect when combined with memoization as reference semantics allow reuse of common subexpression trees.

### CODE REVIEW

**Comment by Sopel97** [comment]
Currently there are a few problems that we need to solve.

1. How to allow building the expression tree in different scopes, considering that we keep references now?
2. How to make memoization work correctly when `eval` is called multiple times with different inputs?

**Comment by Sopel97** [comment]
I believe the problem 2. is now solved. The solution involves creating a unique ID for each expression evaluation, and cache is evicted when the id doesn't match the id from the last call.

**Comment by Sopel97** [comment]
TODO:
1. check the performance against master
2. check if making all steps use `thread_local` changes performance
2.1. Find a way to "batch" thread_local variables.

We cannot just make it constexpr static because it's not thread safe due to caching. We could just keep it as locals but then we can't reference stuff between functions so we have to keep all computation in one. We can't have structs with `auto` members.

**Comment by Sopel97** [comment]
Making each part of the expression tree threadlocal shows `get_loss` as 3% in the profiler, when compiled with MSVC. I'd say that's reasonable.

**Comment by Sopel97** [comment]
Now I've also added code that returns 0 grad early for constant subexpressions, the get_loss profile shows ~2% time is spent there, which is great.

I also made a rough implementation of `use_wdl`, and while it's finally working correctly it might not be the best loss function. It learns very slowly and doesn't like high evals, probably because the sigmoid is very steep it's not incentivised to go to high scores as it doesn't improve perf% a lot.

**Comment by Sopel97** [comment]
This PR is mostly finished. I'll ensure CI works and train a net with tomorrow to ensure nothing regressed (initial sanity checks show that it's fine). It would be nice to get some feedback on this, as it's quite big and we had to make some "hacky" workarounds to make it easy to use.

**Comment by Sopel97** [comment]
Okay, I'll post some results from training tomorrow, and if there are no unexpected problems this PR will be complete.

**Comment by nodchip** [comment]
If my understanding is correct, this PR introduces autograd to calculate the gradient of a loss function.  The idea is awesome.  This PR makes it easy to introduce new loss functions.

**Comment by Sopel97** [comment]
Yes, exactly. But since the gradient from cross entropy is different then previously hardcoded loss function I have to do some tests first :P

**Comment by nodchip** [comment]
The steps to develop the expression of gradients from cross entropy is described in the following document.
https://denou.jp/tournament2016/img/PR/Hakubishin.pdf
I'm sorry but this is written in Japanese.  I will translate if you don't understand it.

**Comment by Sopel97** [comment]
Ok, so the cross entropy gradient does indeed simplify to the calc_grad we have in master. That's reassuring. The automatically differentiated way won't be able to use this simple form obviously, but it has other advantages.

**Comment by Sopel97** [comment]
I trained 3 nets. I used noob's depth 9 data, which I only now realized is no longer accessible, but this should be reproducible with any data. Each net is trained on 100 million fens with lr 1 with smart_fen_skipping instead of qsearch to reduce the time needed to train the net.

The setup:
```
setoption name SkipLoadingEval value true
setoption name Threads value 3
setoption name Use NNUE value pure
learn targetdir data epochs 1000 batchsize 100000 use_draw_in_training 1 use_draw_in_validation 1 lr 1 lambda 1 eval_limit 9000 nn_batch_size 1000 newbob_decay 0.99 eval_save_interval 25000000 loss_output_interval 1000000 max_grad 0.5 smart_fen_skipping set_recommended_uci_options
ucinewgame
```

Logs:
Baseline - trained with the current master: https://pastebin.com/1mw84JFT
Autograd - trained with this PR: https://pastebin.com/SJcQiSgW
Autograd - trained with this PR with `use_wdl 1`: https://pastebin.com/SQhsQYb7

Results:
```
Score of autograd vs baseline: 407 - 323 - 270  [0.542] 1000
Elo difference: 29.25 +/- 18.43
```
This might be explained by a slightly different scaling of the gradient. In master there is no scaling performed. The real gradient is scaled by `winning_probability_coefficient * kPonanzaConstant` with respect to master, which is about `1.66` times higher.

```
Score of autograd-use_wdl vs baseline: 142 - 286 - 111  [0.366] 539
Elo difference: -95.13 +/- 26.88
```
Now this suggests that the loss based on `use_wdl` is not a good loss function for this particular implementation. But it seems to at least work as the training logs look fine and the strength is better than nothing. It is however a good sign that maybe `use_wdl` should be removed. I'll leave it is as it is in this PR for compatibility, but this matter should be addressed in the future in a separate PR.

I'll also leave the unused noob's grad (`get_loss_noob`) in the code so we have some reference implementation for it and it's easier for people to plug in.

I think this PR is ready for a final review.

**Comment by nodchip** [comment]
The result looks great.  Thank you for your great work!
